### ISSUE TITLE
Refactoring map scripts' object_const_def

### ISSUE THREAD

**Issue by Rangi42** [pret/pokecrystal#755]
Each map scripts defines its object constants on top, and defines the object_events on the bottom. As of rgbds 0.4.1, we could move the `object_const_def` block right next to the `object_event` block (doesn't matter which comes first), without the scripts above complaining about undefined constants. Putting the related definitions right next to each other would make their correspondence obvious and help keep them in sync.

More drastically, we could define each constant as part of the `object_event`, like #748 proposes doing for jump tables. If so, I would prefer for the `object_event` X and Y arguments to still come first, since that's consistent with all other map events and is necessary for Polished Map support. (If the argument order changed, like to put a constant name first, then the macro name would also have to change to be distinguishable.) The downside would be even longer lines for `object_event`s; example:

```
        object_event  1, 14, BLACKTHORNGYM1F_COOLTRAINER_M2, SPRITE_COOLTRAINER_M, SPRITEMOVEDATA_STANDING_DOWN, 0, 0, -1, -1, PAL_NPC_RED, OBJECTTYPE_TRAINER, 3, TrainerCooltrainermPaul, -1
````

**Comment by aaaaaa123456789** [comment]
99% of the time you don't need those constants at all (most maps either don't refer to object IDs or don't use any ID greater than 2), and requiring their definition is just clutter. For the remaining 1%, the constants could go anywhere — perhaps as an optional last parameter to the `object_event` macro. (That might be hard without #754.)

**Comment by Rangi42** [comment]
The basic proposal is to just move all the current `object_const_def`s from the top of the file to the bottom, so they'll be adjacent to their `object_event`s. Maps without any `object_event`s already don't have `object_const_def`s. I do like having them defined for all maps because if someone wants to add a `moveobject` or `disappear` or whatever command, they won't have to remember to define constants corresponding to every `object_event`. Still, they wouldn't be required in any case.

An optional last parameter to `object_event` sounds good: only the objects that need them would have them, and they'd automatically correspond to the right object. And yeah, combining that with #754 would let each `def_object_events` macro start a new `const_def 2`, and each `object_event` increments `const_value` whether or not it defines a constant.

**Comment by Rangi42** [comment]
A sample of optional last parameters for 2 out of 5 objects:

```
	def_object_events
	object_event  7,  4, SPRITE_MOM, SPRITEMOVEDATA_STANDING_LEFT, 0, 0, -1, -1, 0, OBJECTTYPE_SCRIPT, 0, MomScript, EVENT_PLAYERS_HOUSE_MOM_1, PLAYERSHOUSE1F_MOM
	object_event  2,  2, SPRITE_MOM, SPRITEMOVEDATA_STANDING_UP, 0, 0, -1, MORN, 0, OBJECTTYPE_SCRIPT, 0, MomScript, EVENT_PLAYERS_HOUSE_MOM_2
	object_event  7,  4, SPRITE_MOM, SPRITEMOVEDATA_STANDING_LEFT, 0, 0, -1, DAY, 0, OBJECTTYPE_SCRIPT, 0, MomScript, EVENT_PLAYERS_HOUSE_MOM_2
	object_event  0,  2, SPRITE_MOM, SPRITEMOVEDATA_STANDING_UP, 0, 0, -1, NITE, 0, OBJECTTYPE_SCRIPT, 0, MomScript, EVENT_PLAYERS_HOUSE_MOM_2
	object_event  4,  4, SPRITE_POKEFAN_F, SPRITEMOVEDATA_STANDING_RIGHT, 0, 0, -1, -1, PAL_NPC_RED, OBJECTTYPE_SCRIPT, 0, NeighborScript, EVENT_PLAYERS_HOUSE_1F_NEIGHBOR, PLAYERSHOUSE1F_NEIGHBOR
```

I like it despite the line length; @mid-kid ?

**Comment by Rangi42** [comment]
So would it make sense to define object constants with an optional last argument to `object_event`, or should I just move the existing `object_const_def` blocks to the bottom of the file?

**Comment by mid-kid** [comment]
I prefer adding them to the same macro, having a separate list is a pain. That said, I really, really don't like the amount of noise on these lines, regardless of the optional constant, and it makes an already confusing macro (due to the multiple optional `0`s that don't line up and can be easily read over) even more confusing.
That said, i don't really see a better approach...

**Comment by Rangi42** [comment]
Yeah, I definitely agree that they're inconveniently long and hard to edit. Lining up the parameters would help, but would also make them even longer.

I'm not sure if multi-line with backslashes is any easier to read:

```
	def_object_events
	object_event  7,  4, SPRITE_MOM,       SPRITEMOVEDATA_STANDING_LEFT,  0, 0, -1, -1,   0, \
	             OBJECTTYPE_SCRIPT, 0, MomScript, EVENT_PLAYERS_HOUSE_MOM_1, PLAYERSHOUSE1F_MOM
	object_event  2,  2, SPRITE_MOM,       SPRITEMOVEDATA_STANDING_UP,    0, 0, -1, MORN, 0, \
	             OBJECTTYPE_SCRIPT, 0, MomScript, EVENT_PLAYERS_HOUSE_MOM_2
	object_event  7,  4, SPRITE_MOM,       SPRITEMOVEDATA_STANDING_LEFT,  0, 0, -1, DAY,  0, \
	             OBJECTTYPE_SCRIPT, 0, MomScript, EVENT_PLAYERS_HOUSE_MOM_2
	object_event  0,  2, SPRITE_MOM,       SPRITEMOVEDATA_STANDING_UP,    0, 0, -1, NITE, 0, \
	             OBJECTTYPE_SCRIPT, 0, MomScript, EVENT_PLAYERS_HOUSE_MOM_2
	object_event  4,  4, SPRITE_POKEFAN_F, SPRITEMOVEDATA_STANDING_RIGHT, 0, 0, -1, -1,   PAL_NPC_RED, \
	             OBJECTTYPE_SCRIPT, 0, NeighborScript, EVENT_PLAYERS_HOUSE_1F_NEIGHBOR, PLAYERSHOUSE1F_NEIGHBOR
```

**Comment by mid-kid** [comment]
I feel like this macro can't be salvaged without named parameters.

**Comment by FredrIQ** [comment]
What about moving the constant prefixes into the macro definition? So you could have object_event 0, 2, MOM, STANDING_UP, ... etc.

**Comment by aaaaaa123456789** [comment]
> I feel like this macro can't be salvaged without named parameters.

Maybe we should implement those, then?

**Comment by Rangi42** [comment]
For now I'll move the `object_const_def`s after the `def_object_event`s, and this can be the "fix `object_event` readability somehow" issue.

**Comment by Rangi42** [comment]
Another object const possibility: `def_object_events` starts a new constant series, `object_event` increments the counter, but instead of defining constants with an optional last argument to `object_event`, have `object_const` on a separate line above the `object_event`. Example:

```
	def_object_events
	object_const PLAYERSHOUSE1F_MOM1
	object_event  7,  4, SPRITE_MOM, SPRITEMOVEDATA_STANDING_LEFT, 0, 0, -1, -1, 0, OBJECTTYPE_SCRIPT, 0, MomScript, EVENT_PLAYERS_HOUSE_MOM_1
	object_event  2,  2, SPRITE_MOM, SPRITEMOVEDATA_STANDING_UP, 0, 0, -1, MORN, 0, OBJECTTYPE_SCRIPT, 0, MomScript, EVENT_PLAYERS_HOUSE_MOM_2
	object_event  7,  4, SPRITE_MOM, SPRITEMOVEDATA_STANDING_LEFT, 0, 0, -1, DAY, 0, OBJECTTYPE_SCRIPT, 0, MomScript, EVENT_PLAYERS_HOUSE_MOM_2
	object_event  0,  2, SPRITE_MOM, SPRITEMOVEDATA_STANDING_UP, 0, 0, -1, NITE, 0, OBJECTTYPE_SCRIPT, 0, MomScript, EVENT_PLAYERS_HOUSE_MOM_2
	object_const PLAYERSHOUSE1F_POKEFAN_F
	object_event  4,  4, SPRITE_POKEFAN_F, SPRITEMOVEDATA_STANDING_RIGHT, 0, 0, -1, -1, PAL_NPC_RED, OBJECTTYPE_SCRIPT, 0, NeighborScript, EVENT_PLAYERS_HOUSE_1F_NEIGHBOR
```

**Comment by aaaaaa123456789** [comment]
I feel like the macro in general could be cleaned up this way, by splitting attributes into multiple lines and letting optional ones go away...

**Comment by Rangi42** [comment]
Maybe so, but "clearly align object constants with object events, only defining constants for objects that use them" and "change how object events are defined instead of with one long macro" are separate goals. Any opinions on this versus having a separate list of constants after all the object events?

**Comment by aaaaaa123456789** [comment]
As I see it, the former looks like a step towards the latter, and while it does seem like a good idea, it is only so if the end goal is to make object definitions multiline anyway.
